airflow/models/taskinstance.py

```diff
@@ -1204,10 +1204,12 @@ def check_and_change_state_before_execution(
             # For reporting purposes, we report based on 1-indexed,
             # not 0-indexed lists (i.e. Attempt 1 instead of
             # Attempt 0 for the first attempt).
-            # Set the task start date. In case it was re-scheduled use the initial
-            # start date that is recorded in task_reschedule table
+            # Set the task start date.
             self.start_date = timezone.utcnow()
-            if self.state == State.UP_FOR_RESCHEDULE:
+
+            # In case it was re-scheduled use the initial
+            # start date that is recorded in task_reschedule table
+            if hasattr(self.task, 'reschedule') and self.task.reschedule:
```

Comment:
So, one issue with this is, sensors are not the only reschedulable operators.   Any operator can raise `AirflowRescheduleException` and  produce the same behavior.  I don't think we can infer what to do based on  looking at task type or attributes.